packages/govern-core/contracts/pipelines/GovernQueue.sol

```diff
@@ -185,22 +185,31 @@ contract GovernQueue is IERC3000, AdaptativeERC165, IArbitrable, ACL {
      */
     function resolve(ERC3000Data.Container memory _container, uint256 _disputeId) override public returns (bytes32 failureMap, bytes[] memory execResults) {
         bytes32 containerHash = _container.hash();
-        if (queue[containerHash].state == GovernQueueStateLib.State.Challenged) {
-            // will re-enter in `rule`, `rule` will perform state transition depending on ruling
-            IArbitrator(_container.config.resolver).executeRuling(_disputeId);
-        } // else continue, as we must 
+        IArbitrator arbitrator = IArbitrator(_container.config.resolver);
+
+        require(disputeItemCache[containerHash][arbitrator] == _disputeId, "queue: bad dispute id");
+        delete disputeItemCache[containerHash][arbitrator];
 
-        GovernQueueStateLib.State state = queue[containerHash].state;
+        (address subject, uint256 ruling) = arbitrator.rule(_disputeId);
+        require(subject == address(this), "queue: not subject");
+        bool arbitratorApproved = ruling == ALLOW_RULING;
 
-        emit Resolved(containerHash, msg.sender, state == GovernQueueStateLib.State.Approved);
+        GovernQueueStateLib.State newState = arbitratorApproved
+            ? GovernQueueStateLib.State.Approved
+            : GovernQueueStateLib.State.Rejected;
+        queue[containerHash].checkAndSetState(
+            GovernQueueStateLib.State.Challenged, // checks now that this container was indeed in Challenged state
+            newState
+        );
 
-        if (state == GovernQueueStateLib.State.Approved) {
+        emit Resolved(containerHash, msg.sender, arbitratorApproved);
+        emit Ruled(arbitrator, _disputeId, ruling);
```

Comment:
It seems a bit weird we have two different events for this, but not sure if there's anything we can do.

On the protocol side, when we considered the `Ruled` event, it was only to set a standard event so that it'd be easier to find Arbitrables later from the Protocol's side (which we assume will be blind to ERC3k's interfaces / internals).